vllm/config.py

```diff
@@ -680,26 +680,32 @@ def get_num_attention_heads(self,
         num_heads = getattr(self.hf_text_config, "num_attention_heads", 0)
         return num_heads // parallel_config.tensor_parallel_size
 
-    def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
+    def get_layers_start_end_indices(
+            self, parallel_config: "ParallelConfig") -> Tuple[int, int]:
         from vllm.distributed.utils import get_pp_indices
         total_num_hidden_layers = getattr(self.hf_text_config,
                                           "num_hidden_layers", 0)
         pp_rank = parallel_config.rank // parallel_config.tensor_parallel_size
         pp_size = parallel_config.pipeline_parallel_size
         start, end = get_pp_indices(total_num_hidden_layers, pp_rank, pp_size)
-        return end - start
+        return start, end
 
-    def get_num_attention_layers(self,
-                                 parallel_config: "ParallelConfig") -> int:
-        if self.is_attention_free:
-            return 0
+    def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
+        start, end = self.get_layers_start_end_indices(parallel_config)
+        return end - start
 
-        num_layers = self.get_num_layers(parallel_config)
+    def get_num_layers_by_block_type(
+        self,
+        parallel_config: "ParallelConfig",
+        block_type: LayerBlockType = LayerBlockType.attention,
+    ) -> int:
+        start, end = self.get_layers_start_end_indices(parallel_config)
 
         # Transformers supports layers_block_type @property
-        layers = getattr(self.hf_config, "layers_block_type",
-                         ["attention"] * num_layers)
-        return len([t for t in layers if t == "attention"])
+        layers_block_type_value = getattr(self.hf_config, "layers_block_type",
+                                          [block_type.value] * (end - start))
+        return sum(t == block_type.value
+                   for t in layers_block_type_value[start:end])
```

Comment:
@mzusman do you know if this function is well-behaved for `mamba`?

Reply:
I just checked and there's a problem here: If there is no `layers_block_type` in `self.hf_config`, then this function will always return `num_layers`. This happens for `mamba` at least

Reply:
@tlrmchlsmth it is well-behaved for mamba, I've run mamba_test locally and it seems to behave fine (except for the multistep scheduling tests - It's unrelated, will take care of it in a future PR). 
I've also changed the mamba cache creation in the modeling file to use this function instead of the config.num_hidden_layers so in each block in the PP will not allocate the whole cache.
Are there other concerns?

Reply:
@tlrmchlsmth You're right, I've missed that, please see my fix

Reply:
There's still a problem where  
```
model_config.get_num_layers_by_block_type(parallel_config, LayerBlockType.mamba)
```
will return  `num_layers` if there is no `layers_block_type` attribute. This is the case for llama for instance.

Unfortunately I don't think there's a standard attribute in the huggingface configs that we can query. One option would be to add an `IsHybridModel` `Protocol` in `model_executor/models/interfaces.py`, using it in Jamba and checking it here like we did for `is_attention_free`

Reply:
At least Zamba has a `layers_block_type` (https://huggingface.co/Zyphra/Zamba-7B-v1) so hopefully hybrid models end up standardizing on that

Reply:
Changed it accordingly. WDYT?